Sources/ConnectionPoolModule/PoolStateMachine.swift

```diff
@@ -374,6 +374,14 @@ struct PoolStateMachine<
         return self.handleAvailableConnection(index: index, availableContext: context)
     }
 
+    @inlinable
+    mutating func connectionKeepAliveFailed(_ connection: Connection) -> Action {
+        if self.connections.keepAliveFailed(connection.id) {
+            return self.connectionClosed(connection)
```

Comment:
That doesn‘t look right, if we the keep alive failed, we can‘t declare the connection dead right away, we need to create a close action. Also if we close and have waiting requests, we might need to create a new connection right away.

Reply:
The connection is now properly closed by reusing the existing closed states and actions. I've added a test to ensure a new connection is established if required.